store/postgres/src/relational_queries.rs

```diff
@@ -2008,37 +2023,84 @@ impl<'a, Conn> RunQueryDsl<Conn> for FindQuery<'a> {}
+            is_upper_range,
+            imm_range,
+            mut_range,
+        }
     }
 }
 
 impl<'a> QueryFragment<Pg> for FindRangeQuery<'a> {
     fn walk_ast<'b>(&'b self, mut out: AstPass<'_, 'b, Pg>) -> QueryResult<()> {
         out.unsafe_to_cache_prepared();
+        let mut first = true;
 
-        // Generate
-        //    select '..' as entity, to_jsonb(e.*) as data
-        //      from schema.table e where id = $1
-        out.push_sql("select ");
-        out.push_bind_param::<Text, _>(self.table.object.as_str())?;
-        out.push_sql(" as entity, to_jsonb(e.*) as data\n");
-        out.push_sql("  from ");
-        out.push_sql(self.table.qualified_name.as_str());
-        out.push_sql(" e\n where ");
-        // TODO: do we need to care about it?
-        // if self.table.has_causality_region {
-        //     out.push_sql("causality_region = ");
-        //     out.push_bind_param::<Integer, _>(&self.key.causality_region)?;
-        //     out.push_sql(" and ");
-        // }
-        self.eb_range.contains(&mut out)
+        for table in self.tables.iter() {
+            // the immutable entities don't have upper range and also can't be modified or deleted
+            if !(self.is_upper_range && table.immutable) {
+                if first {
+                    first = false;
+                } else {
+                    out.push_sql("\nunion all\n");
+                }
+
+                // Generate
+                //    select '..' as entity, to_jsonb(e.*) as data, block$ as block_number
+                //      from schema.table e where id = $1
+                out.push_sql("select ");
+                out.push_bind_param::<Text, _>(table.object.as_str())?;
+                out.push_sql(" as entity, to_jsonb(e.*) as data,");
+                if table.immutable {
+                    self.imm_range.compare_column(&mut out)
+                } else {
+                    self.mut_range.compare_column(&mut out)
+                }
+                out.push_sql("as block_number, id, vid\n");
+                out.push_sql("  from ");
+                out.push_sql(table.qualified_name.as_str());
+                out.push_sql(" e\n  where");
+                // add casuality region to the query
+                if table.has_causality_region {
+                    out.push_sql("causality_region = ");
+                    out.push_bind_param::<Integer, _>(&self.causality_region)?;
+                    out.push_sql(" and ");
+                }
+                if table.immutable {
+                    self.imm_range.contains(&mut out)?;
+                } else {
+                    self.mut_range.contains(&mut out)?;
+                }
+            }
+        }
+
+        if first {
+            // In case we have only immutable entities, the upper range will not create any
+            // select statement. So here we have to generate an empty SQL statement.
+            out.push_sql("select 1");
```

Comment:
Does this work? If we have clauses in the query, it returns `(text, jsonb, int)`, but this fallback just returns `(int)` and is therefore the wrong shape

Reply:
I believe I tested it. The intention was to return an empty set. Now when I run it on the command line it returns one row, so maybe diesel silently substituted wrong types with empty rows, but its a bad style for sure. Needs fixing.

Reply:
[Addressed](https://github.com/graphprotocol/graph-node/pull/5614/commits/5e4df3863532eceb8316c3ac28b148c86aeef34e) it and added a test for that case.